common/djangoapps/util/db.py

```diff
@@ -4,44 +4,213 @@
-                elif len(connection.transaction_state) > 1:
-                    raise transaction.TransactionManagementError('Cannot change isolation level. '
-                                                                 'More than 1 level of nested transactions.')
+    # Tests in TestCase subclasses are wrapped in an atomic block to speed up database restoration.
+    # So we must disabled this manager.
+    # https://github.com/django/django/blob/1.8.5/django/core/handlers/base.py#L129-L132
+    ENABLED = True
+
+    def __init__(self, using, read_committed=False):
+        self.using = using
+        self.read_committed = read_committed
+
+    def __enter__(self):
+
+        if not self.ENABLED:
+            return
+
+        connection = transaction.get_connection(self.using)
+
+        if connection.in_atomic_block:
+            raise transaction.TransactionManagementError('Cannot be inside an atomic block.')
+
+        if getattr(connection, 'commit_on_success_block_level', 0) == 0:
+            connection.commit_on_success_block_level = 1
 
             # This will set the transaction isolation level to READ COMMITTED for the next transaction.
-            cursor = connection.cursor()
-            cursor.execute("SET TRANSACTION ISOLATION LEVEL READ COMMITTED")
+            if self.read_committed is True:
+                if connection.vendor == 'mysql':
+                    cursor = connection.cursor()
+                    cursor.execute("SET TRANSACTION ISOLATION LEVEL READ COMMITTED")
+
+            # We aren't in a transaction yet; create one.
+            # The usual way to start a transaction is to turn autocommit off.
+            # However, some database adapters (namely sqlite3) don't handle
+            # transactions and savepoints properly when autocommit is off.
+            # In such cases, start an explicit transaction instead, which has
+            # the side-effect of disabling autocommit.
+            if connection.features.autocommits_when_autocommit_is_off:
+                connection._start_transaction_under_autocommit()
+                connection.autocommit = False
+            else:
+                connection.set_autocommit(False)
+        else:
+            if self.read_committed is True:
+                raise transaction.TransactionManagementError('Cannot change isolation level when nested.')
+
+            connection.commit_on_success_block_level += 1
+
+    def __exit__(self, exc_type, exc_value, traceback):
+
+        if not self.ENABLED:
+            return
+
+        connection = transaction.get_connection(self.using)
+
+        try:
+            if exc_type is None:
+                # Commit transaction
+                try:
+                    connection.commit()
+                except DatabaseError:
+                    try:
+                        connection.rollback()
+                    except Error:
+                        # An error during rollback means that something
```

Comment:
I'll put in some logging statements here. Though ideally we want to get rid of this `commit_on_success` decorator before we deploy the upgrade. Right now it is only being used in `courseware/grades.py` and on the progress page view. I have created a separate story for it.